certora/specs/Splits.spec

```diff
@@ -0,0 +1,418 @@
+/// We also verify that different inputs must generate different hashes
+rule correctnessOfHashSplits() {
+    env e; uint256 index; uint256 length1; uint256 length2;  
+    uint256 userId1; uint32 weight1; uint256 userId2; uint32 weight2;
+    bytes32 receiversHash1; bytes32 receiversHash2;
+
+    length1 = getCurrSplitsReceiverLocaLength(e, true);
+    length2 = getCurrSplitsReceiverLocaLength(e, false);
+
+    userId1, weight1 = getCurrSplitsReceiverLocalArr(e, true, index);
+    userId2, weight2 = getCurrSplitsReceiverLocalArr(e, false, index);
+
+    receiversHash1 = hashSplits(e, true);
+    receiversHash2 = hashSplits(e, false);
+
+    assert (receiversHash1 == receiversHash2) => (length1 == length2);
+    assert (receiversHash1 == receiversHash2) => (userId1 == userId2);
+    assert (receiversHash1 == receiversHash2) => (weight1 == weight2);
+
+    assert (length1 != length2) => (receiversHash1 != receiversHash2);
+    assert (length1 == length2) && (userId1 != userId2) => (receiversHash1 != receiversHash2);
+    assert (length1 == length2) && (weight1 != weight2) => (receiversHash1 != receiversHash2);
+
+    // assert (receiversHash1 == receiversHash2) <=> ((length1 == length2) && (userId1 == userId2) && (weight1 == weight2));
+    // check with @Roy the commented line above
+}
+
+
+/// @notice Integrity of split:
+/// UserA has received drips and he has splittable > 0
+/// UserA has a list of splitters that should get some of the drips received
+/// UserA calls split()
+/// UserB is on the list of UserA's splitters, therefore "userB's splittable should NOT decrease"
+/// UserC is NOT on the list of UserA's splitters, therefore "userC's splittable should NOT change"
+/// UserA's collectable should NOT decrease, UserB and UserC's collectable should NOT change
+rule integrityOfSplit() {
+    env e; uint256 assetId;
+    uint256 userA_Id; uint256 userB_Id; uint256 userC_Id;
+    
+    require userA_Id != userB_Id;
+    require userA_Id != userC_Id;
+    require userB_Id != userC_Id;
+
+    // obtaining splittable/collectable states before calling split()
+    uint128 userA_splittableBefore = splittable(e, userA_Id, assetId);
+    require userA_splittableBefore > 0;
+    uint128 userA_collectableBefore = collectable(e, userA_Id, assetId);
+    uint128 userB_splittableBefore = splittable(e, userB_Id, assetId);
+    uint128 userB_collectableBefore = collectable(e, userB_Id, assetId);
+    uint128 userC_splittableBefore = splittable(e, userC_Id, assetId);
+    uint128 userC_collectableBefore = collectable(e, userC_Id, assetId);
+
+    // setting up the currReceivers[] of userA
+    uint256 index1; uint256 userId1; uint32 weight1;
+    uint256 index2; uint256 userId2; uint32 weight2;
+    userId1, weight1 = getCurrSplitsReceiverLocalArr(e, true, index1);
+    userId2, weight2 = getCurrSplitsReceiverLocalArr(e, true, index2);
+
+    require index1 != index2;
```

Comment:
Explain these 4 require statements better. What state are you trying to force here?

Reply:
I added the following explainer:
```
    require index1 != index2;     // different indexes sample different splitReceivers
    require userId1 == userB_Id;  // userB is on the list of splitReceivers of userA
    require userId2 != userC_Id;  // the second splitReceiver is not userC
    require userId2 != userA_Id;  // the second splitReceiver is not the splitter itself
                                  // since we run with loop_iter 2, there are max 2 receivers
```